### PR TITLE
Fix mods potentially adjusting track while not selected

### LINKED ISSUE #18178: Deselecting Wind Up/Down mod when pitch adjustment is turned off results in it turning back on
### Type

Game behaviour

### Bug description

simple to reproduce simply select the wind up/down mod then turn off the pitch adjustment and deselect the mod

### Screenshots or videos

https://user-images.githubusercontent.com/43358824/167393318-b26d97b6-f3c3-4eb6-9b7d-6c70a13ca051.mp4

### Version

2022.509.0

### Logs

[database.log](https://github.com/ppy/osu/files/8650888/database.log)
[network.log](https://github.com/ppy/osu/files/8650889/network.log)
[performance.log](https://github.com/ppy/osu/files/8650890/performance.log)
[runtime.log](https://github.com/ppy/osu/files/8650891/runtime.log)
[updater.log](https://github.com/ppy/osu/files/8650892/updater.log)

### PR DESCRIPTION
**Pull request by frenzibyte**
- Closes #18178 
- Alternative to (but doesn't close) #18196 

The issue is better described in #18196, and this solves it alternatively by isolating mod adjustments from `CurrentTrack` using an `AudioAdjustments` layer, to allow for `MusicController` to completely unbind from these adjustments when no mod is selected. Rather than passing the track directly and have it kept as a reference for mods like `ModTimeRamp`.

Importantly, and as I was looking at the changes here and in the other PR, it's notable to say that this PR only achieves in avoiding mods from potentially applying adjustments to the track while they are not selected or otherwise. But such mods should never store the track and use it outside of scope like that, and I think #18196 solves that pretty well by local bindables instead.

Therefore I would say this is only an optional improvement PR which can either be merged or closed, but the `ModTimeRamp`/`ModAdaptiveSpeed` changes in #18196 feel much better in general, as they avoid the necessity of storing a track in the first place.

Also, since `ITrack`s are no longer passed to the mods, `ModTimeRamp` has been changed to figure out the current time using `playfield.Clock.CurrentTime` instead, which... comes with this funny side effect:   

https://user-images.githubusercontent.com/22781491/167667792-7b73c354-0150-4443-8fbf-23375d261bb6.mp4

### CODE REVIEW

File: osu.Game/Overlays/MusicController.cs
Comment (bdach) [inline osu.Game/Overlays/MusicController.cs:404]: I see that this has changed somewhat from the original diff you posted (and is now more correct), but my gripe here is thus: how is this any safer than the previous code? Any mod is still free to hold references to `modTrackAdjustments` and mess with them after it's not supposed to, because it's still a single shared reference.

If we're so insistent on using the adjustment layer, then why doesn't this method just create a *new* `AudioAdjustments` every time it's called, discard the previous one, and expose the new temporary one to the mods? Which is to say:

```csharp
private AudioAdjustments modAdjustments;

public void ResetTrackAdjustments()
{
    // remove everything from CurrentTrack as in the original

    CurrentTrack.UnbindAdjustments(modAdjustments);

    // statement below equivalent to removing all adjustments as done in your version
    CurrentTrack.BindAdjustments(modAdjustments = new AudioAdjustments());

    foreach (var mod in applicableToTrack)
        mod.ApplyToTrack(modAdjustments);
}
```

Note that this *ensures* that mods cannot ever do anything bad with the `AudioAdjustments` they're given, because the references they're given are thrown away after each `ResetTrackAdjustments()` call.

File: osu.Game/Rulesets/Mods/ModTimeRamp.cs
Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/ModTimeRamp.cs:91]: Well, that's certainly no longer funny... definitely needs an inline comment mentioning that to avoid potentially changing it in the future, very spooky.

Not entirely sure how to solve this one, but one idea I can think of is providing the gameplay clock as part of `Playfield` and expose the source track clock from there.

Comment (bdach) [reply osu.Game/Rulesets/Mods/ModTimeRamp.cs:91]: I personally wouldn't go there. You could change signature of `ApplyToTrack()` to something like

```csharp
void ApplyToTrack(IClock track, IAdjustableAudioComponent adjustments);
```

Or even just cast it once in `ModTimeRamp` and call it a day. I don't think exposing more stuff in the already-too-complex gameplay clock machinery serves anyone any good, but I'll wait and see what everyone else has to say about it.

Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/ModTimeRamp.cs:91]: I would avoid storing a reference to yet another track in `Mod` at all costs IMO, and would want to completely decouple the idea of track consumption from `ApplyToTrack` which was only supposed to be for adding adjustments to the audio component at hand.

Comment (bdach) [reply osu.Game/Rulesets/Mods/ModTimeRamp.cs:91]: I think that if exposed by `IClock` it's fine. You can't do anything bad with an `IClock`, unless you cast, at which point you've firmly blown your foot off on your own because that's not how interfaces work and cannot realistically complain that things are not working right.

Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/ModTimeRamp.cs:91]: Maybe? For me at first, thinking about `ApplyToTrack` storing a track `IClock` reference inside and outside gameplay context yet being consumed in `Update(Playfield)` only kinda broke my mind there, as opposed to letting `Playfield` and `GameplayClock` expose the source time for the mod. Will wait for other opinions.

Comment (smoogipoo) [reply osu.Game/Rulesets/Mods/ModTimeRamp.cs:91]: I'm not entirely sure that the track clock is correct to use over the `FrameStableClock` specifically. Because in the end, it's that clock which determines the point of time which gameplay is at, the track clock is almost like a hint.

Don't we refer to the FrameStableClock in other mods?

Also thoughts @peppy?

Comment (peppy) [reply osu.Game/Rulesets/Mods/ModTimeRamp.cs:91]: Thinking about it in terms of ramping is probably beneficial. Right now, when doing a seek, the audio seeks immediately (though I'd argue it shouldn't). If switching to providing the `FrameStableClock` here, the seek will be immediate but the ramp would be applied as the game catches up. It might feel a bit odd?

Comment (smoogipoo) [reply osu.Game/Rulesets/Mods/ModTimeRamp.cs:91]: I don't think this affects catchup? Because it's affecting the progression of time rather than the length of time, if that makes sense. Put another way, it affects how long a millisecond is _compared to real time_ but `T=1000` is always the same point in time no matter how it's reached.  
Catchup always happens at max ~80ms intervals (16ms frame stability * 5 update frames) independent of real-time.

Rather I was thinking about it from the perspective of offsets. I'm not sure if I'm correct since this is kinda hard to mentally parse, but if rate is applied relative to track time then it's possible that two systems with different offsets will have completely different playback.  
Suppose an arbitrary (either UO or local) offset of -1000ms meaning that game time is 1000ms behind track time (0ms track time = -1000ms gameplay time), and suppose that there's a hitobject at T=0. By the time gameplay hits T=0, the track time is going to be at T=1000ms and the approach circle for that hitobject will experience an increased time ramp compared to an offset of 0ms.

Comment (peppy) [reply osu.Game/Rulesets/Mods/ModTimeRamp.cs:91]: For application of rate adjusts, I think applying to the "source of truth" is best. I think that's the track in this case. As long as we don't reverse that then the track's clock is best.

Correct me if I'm wrong, but the discussion here is proposing potentially reading the currrent time from `FrameStableClock` for use in
 https://github.com/ppy/osu/blob/8501a4161958514bfae52c5e78aee2ee26a870e0/osu.Game/Rulesets/Mods/ModTimeRamp.cs#L93

right?

If so, I think what I said is correct in that the rate applied will change during seeks.

Comment (bdach) [reply osu.Game/Rulesets/Mods/ModTimeRamp.cs:91]: The frame stable clock _without_ offsets _might_ (and I'm saying this very carefully) be more correct than the track clock in one scenario specifically, which is seeking during replay playback. I've touched on that [in the past](https://github.com/ppy/osu/pull/11697#issuecomment-775733845), however I'm still pretty sure that any clock that takes user offsets into effect is not valid to use with `ModTimeRamp`.

Comment (frenzibyte) [reply osu.Game/Rulesets/Mods/ModTimeRamp.cs:91]: I’ve had thought about this for a bit and seeing what @smoogipoo has mentioned already, isn’t using `FrameStableClock` with offsets actually correct? 

What you want to be ramped is the gameplay clock, so that the speed starts changing as you progress through the hitobjects, constantly, regardless of any offset applied, which is what `GameplayClock/`FrameStableClock` should preserve.

When ramping with the source track clock, the speed starts change while you haven’t reached the first object yet, because the objects are offset from the track time.

Comment (peppy) [reply osu.Game/Rulesets/Mods/ModTimeRamp.cs:91]: For gameplay, frame stable is the correct clock to use in all cases. I was focusing on the replay case because that's one area that will be less correct if changing this flow to use `FrameStableClock`.

But I think it's more correct to use it from a gameplay perspective, so I'd be in favour of that for now.

Comment (bdach) [reply osu.Game/Rulesets/Mods/ModTimeRamp.cs:91]: Aside from the fact that now we're arguing on clock minutiae on a pull request supposing to fix an entirely unrelated regression, I acknowledge that using the track as a clock does potentially lead to a scenario wherein i.e. approach circles move slower or faster depending on offset, I still believe applying rate based on clocks with offset to be incorrect, because if you use that, then you actually ever so slightly modify the range of rate ramping (because at track time 0, frame stable clock w/offsets can be anywhere from -500 to +500ms). That seems more incorrect to me than visuals shifting.

That said, I'm mostly done with this conversation and I'll wait for everyone to make the final decision.

Comment (peppy) [reply osu.Game/Rulesets/Mods/ModTimeRamp.cs:91]: To confirm we're on the same page: `FrameStableClock` has the offsets applied, so you are okay with using this, correct?

Comment (bdach) [reply osu.Game/Rulesets/Mods/ModTimeRamp.cs:91]: I don't think that is correct, no, my claim remains that user offsets should not affect time ramp. But that's more or less a hunch based on the argument that the offset being applied is changing initial track rate, and the clock workings are so complex that I can't reasonably say whether I'm correct in this assertion.

Comment (peppy) [reply osu.Game/Rulesets/Mods/ModTimeRamp.cs:91]: That means that two users with different user offsets would have different rates at the same hitobject, no? Seems wrong from a gameplay correctness/fairness perspective.

Comment (bdach) [reply osu.Game/Rulesets/Mods/ModTimeRamp.cs:91]: I actually don't think either option is fair anymore. Either the visuals are sped up/slowed down depending on offset, or the track is. It feels more correct to me to leave the track in place as it's closer to being the "source" of time.

Comment (smoogipoo) [reply osu.Game/Rulesets/Mods/ModTimeRamp.cs:91]: @bdach Just to clarify, your case is when someone is abusing offsets, correct? I wasn't considering that case, only in the case where an offset is used to correct hitsound/gameplay <-> audio time, so for a non-malicious reason.

Comment (bdach) [reply osu.Game/Rulesets/Mods/ModTimeRamp.cs:91]: Yeah as mentioned on discord seconds ago, if the offset application is to fix alignment with hitobjects, then I can agree that using post offset clock is correct.

File: osu.Game/Rulesets/Mods/IApplicableToTrack.cs
Comment (peppy) [inline osu.Game/Rulesets/Mods/IApplicableToTrack.cs:13]: Wonder how many rulesets this will break (if any?)

File: osu.Game/Rulesets/Mods/ModTimeRamp.cs
Comment (peppy) [reply osu.Game/Rulesets/Mods/ModTimeRamp.cs:91]: I'd much prefer this was a direct access to a `GameplayClock`, but I guess it's fine as-is.

**Review by peppy — APPROVED** [review]

**Comment by smoogipoo** [comment]
Why does this not close #18196?

**Comment by frenzibyte** [comment]
For me, the changes in #18196 still look valid (https://github.com/ppy/osu/pull/18196/commits/a54295b31baf4ee742fc7bd75c884f242b94a696 specifically) in that it avoids storing an audio component reference by replacing it with local bindables instead, so I didn't mark this PR as fully alternative and closes it. But if you think one should replace the other, then you can go ahead.

**Comment by bdach** [comment]
If this is merged then I will be closing mine regardless.

File: osu.Game/Overlays/MusicController.cs
Comment (smoogipoo) [inline osu.Game/Overlays/MusicController.cs:401]: It feels like a half-solution when the OP is saying the changes in #18196 feel better (referring to applying via bindables/not storing the "track") but this is still feeding into that.

Have you considered changing these methods to `CreateTrackAdjustments`/`CreateSampleAdjustments()` instead, to make the mods return `AudioAdjustment` objects?

Comment (bdach) [reply osu.Game/Overlays/MusicController.cs:401]: The adjustments that music controller feeds to mods are thrown away on every adjustment reset. So it no longer matters what the mods do, because they essentially can only store temporary adjustments even if they try to store them.

File: osu.Game/Overlays/MusicController.cs
Comment (smoogipoo) [inline osu.Game/Overlays/MusicController.cs:394]: This could now probably be replaced with:
```
if (modTrackAdjustments != null)
{
    CurrentTrack.UnbindAdjustments(modTrackAdjustments);
    modTrackAdjustments = null;
}
```

**Review by smoogipoo — DISMISSED** [review]

File: osu.Game/Overlays/MusicController.cs
Comment (frenzibyte) [reply osu.Game/Overlays/MusicController.cs:394]: Well that's what I had hoped for, but `CurrentTrack` could have adjustments applied directly from other screens like `Player`. Your other suggestion in https://github.com/ppy/osu/pull/18211#discussion_r870983474 may actually work best (defining an `AudioAdjustments` property exposed in `Mod` and lets usages bind to it), but maybe for another day?

**Comment by smoogipoo** [comment]
This doesn't look to be working:
1. Apply Wind Up mod.
2. Complete gameplay.
3. Deselect mod.
4. Mod adjustments are still applied.

At this point, if you apply wind up again, it will double up. I believe it doubles up even if you don't deselect the mod in (3).

https://user-images.githubusercontent.com/1329837/168007312-8c01ba5c-d8f7-42c1-a2f1-def524f99f04.mp4

**Review by smoogipoo — CHANGES_REQUESTED** [review]
As mentioned

**Comment by smoogipoo** [comment]
Nevermind, I had local changes >_<

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**
